File: packages/guards/src/passport-detect.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface PassportDetectOptions {
  action: 'block' | 'warn';
}

const PASSPORT_PATTERNS: RegExp[] = [
  /\b[A-Z]\d{8}\b/,              // US (1 letter + 8 digits)
  /\b\d{9}\b/,                    // US numeric only
  /\b[0-9]{9}[A-Z]{2}\b/,        // UK old format
  /\b[A-Z]{2}\d{7}\b/,           // EU common (2 letters + 7 digits)
  /\b[A-Z]{1}\d{7}[A-Z]{1}\b/,   // some EU
  /\b[MG]\d{8}\b/,                // Korean (M/G + 8 digits)
  /\b[A-Z]{2}\d{7}\b/,           // Japanese (2 letters + 7 digits)
  /\b[EG]\d{8}\b/,                // Chinese (E/G + 8 digits)
];

export function passportDetect(options: PassportDetectOptions): Guard {
  return {
    name: 'passport-detect',
    version: '0.1.0',
    description: 'Detects international passport numbers',
    category: 'privacy',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const pattern of PASSPORT_PATTERNS) {
        if (pattern.test(text)) {
          matched.push(pattern.source);
        }
      }

      const triggered = matched.length > 0;

      return {
        guardName: 'passport-detect',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score: triggered ? Math.min(matched.length / 2, 1.0) : 0,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matchedPatterns: matched.length } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/passport_detect.py
"""Detect international passport numbers."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_PASSPORT_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"\b[A-Z]\d{8}\b"),
    re.compile(r"\b[A-Z]{2}\d{7}\b"),
    re.compile(r"\b[A-Z]{1}\d{7}[A-Z]{1}\b"),
    re.compile(r"\b[MG]\d{8}\b"),
    re.compile(r"\b[EG]\d{8}\b"),
    re.compile(r"\b\d{9}[A-Z]{2}\b"),
]


class _PassportDetect:
    def __init__(
        self, *, action: str = "block"
    ) -> None:
        self.name = "passport-detect"
        self.action = action

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []

        for pat in _PASSPORT_PATTERNS:
            if pat.search(text):
                matched.append(pat.pattern)

        triggered = len(matched) > 0
        score = (
            min(len(matched) / 2, 1.0)
            if triggered
            else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="passport-detect",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Passport number detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"matched_patterns": len(matched)}
                if triggered
                else None
            ),
        )


def passport_detect(
    *, action: str = "block"
) -> _PassportDetect:
    return _PassportDetect(action=action)
